<?xml version="1.0" encoding="utf-8"?>
<!-- This URDF was automatically created by SolidWorks to URDF Exporter! Originally created by Stephen Brawner (brawner@gmail.com) 
     Commit Version: 1.6.0-1-g15f4949  Build Version: 1.6.7594.29634
     For more information, please see http://wiki.ros.org/sw_urdf_exporter -->

<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="cr3_robot">

  <!-- xacro macro-->
  <xacro:macro name="arm_gazebo" params="link">
            <gazebo reference="${link}">
                <material>Gazebo/White</material>
                <mu1>0.2</mu1>
                <mu2>0.2</mu2>
            </gazebo>
  </xacro:macro>

  <xacro:macro name="arm_control" params="transmission joint motor">
    <transmission name="${transmission}">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="${joint}">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
      <actuator name="${motor}">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>
  </xacro:macro>
  <!--  -->

  <link name="world" />
  <link name="base_link">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="3.06" />
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 -1.57" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/base_link.dae" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/base_link.STL" />
      </geometry>
    </collision>
  </link>

  <joint name="world_joint" type="fixed">
    <parent link="world" />
    <child link="base_link" />
  </joint>

  <link name="Link1">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="3.06" />
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link1.dae" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link1.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="joint1" type="revolute">
    <origin xyz="0 0 0.1283" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="Link1" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14" upper="3.14" effort="0" velocity="0" />
    <dynamics damping="1.0" />
  </joint>
  <link name="Link2">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="3.06" />
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link2.dae" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link2.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="joint2" type="revolute">
    <origin xyz="0 0 0" rpy="1.5708 1.5708 0" />
    <parent link="Link1" />
    <child link="Link2" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14" upper="3.14" effort="0" velocity="0" />
    <dynamics damping="1.0" />
  </joint>
  <link name="Link3">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="3.06" />
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link3.dae" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link3.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="joint3" type="revolute">
    <origin xyz="-0.274 0 0" rpy="0 0 0" />
    <parent link="Link2" />
    <child link="Link3" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14" upper="3.14" effort="0" velocity="0" />
    <dynamics damping="1.0" />
  </joint>
  <link name="Link4">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="3.06" />
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link4.dae" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link4.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="joint4" type="revolute">
    <origin xyz="-0.23 0 0.1283" rpy="0 0 -1.5708" />
    <parent link="Link3" />
    <child link="Link4" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14" upper="3.14" effort="0" velocity="0" />
    <dynamics damping="1.0" />
  </joint>
  <link name="Link5">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="3.06" />
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link5.dae" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link5.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="joint5" type="revolute">
    <origin xyz="0 -0.116 0" rpy="1.5708 0 0" />
    <parent link="Link4" />
    <child link="Link5" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14" upper="3.14" effort="0" velocity="0" />
    <dynamics damping="1.0" />
  </joint>
  <link name="Link6">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="3.06" />
      <inertia ixx="0.3" ixy="0.0" ixz="0.0" iyy="0.3" iyz="0.0" izz="0.3" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link6.dae" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://cr3_lowlevel_pkg/meshes/Link6.STL" />
      </geometry>
    </collision>
  </link>
  <joint name="joint6" type="revolute">
    <origin xyz="0 0.105 0" rpy="-1.5708 0 0" />
    <parent link="Link5" />
    <child link="Link6" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14" upper="3.14" effort="0" velocity="0" />
    <dynamics damping="1.0" />
  </joint>

  <!-- xacro implement-->
  
  <xacro:arm_control transmission="tran_1" joint="joint1" motor="motor_1" />
  <xacro:arm_control transmission="tran_2" joint="joint2" motor="motor_2" />
  <xacro:arm_control transmission="tran_3" joint="joint3" motor="motor_3" />
  <xacro:arm_control transmission="tran_4" joint="joint4" motor="motor_4" />
  <xacro:arm_control transmission="tran_5" joint="joint5" motor="motor_5" />
  <xacro:arm_control transmission="tran_6" joint="joint6" motor="motor_6" />
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so" />
  </gazebo>

  <xacro:arm_gazebo link="link0"/>
        <xacro:arm_gazebo link="link1"/>
        <xacro:arm_gazebo link="link2"/>
        <xacro:arm_gazebo link="link3"/>
        <xacro:arm_gazebo link="link4"/>
        <xacro:arm_gazebo link="link5"/>
        <xacro:arm_gazebo link="link6"/>

</robot>